_posts/2016-02-25-on-idempotency-in-distributed-rest-apis.md

```diff
@@ -0,0 +1,132 @@
+``GET``, ``PUT``, ``POST`` and ``DELETE``.
+
+| Method | Safe | Idempotent |
+|:-------|:----:|:----------:|
+| GET    | Yes  | Yes        |
+| PUT    | No   | Yes        |
+| POST   | No   | No         |
+| DELETE | No   | Yes        |
+
+
+## Pragmatic vs Idempotency
+
+Being idempotent is important. However, it does not mean that you
+have to guarantee that a request one point
+in time should always return the same result. E.g. applying ``DELETE`` on
+a existing resource multiple times should return ``HTTP Status: OK (200)``
+the first time and ``HTTP Status: Not Found (404)`` the subsequent times.
+
+A strict implementation of ``DELETE`` would expect ``HTTP Status: OK (200)``.
+I believe that approach to be unnecessary cumbersome and pragmatism should
+trump idempotency in such cases.
+
+
+## Concurrency vs Idempotency
+
+A typical semantic used to ensure idempotency for a ``PUT`` request is to
+require the client to send all the values, including the ones
+that does not change, when updating an object. This approach results in an
+``PUT`` that is idempotent but prone to race conditions, both in case of
+network failure and in highly concurrent environments.
+
+Say ``Alice`` tries to update the ``secret`` to ``"A"`` and because the network
+is not reliable she loses the connection right before the request was supposed
+to respond. Now she does not know if the the update got through or not. However,
+since the ``PUT`` is idempotent she can retry the request without worrying of
+causing other side effects.
+
+This line of thought is correct, if we assume that ``Alice`` was the only only one
+trying to update the secret. Let us assume that an another user Bob
+was simultaneously trying to update the same ``secret`` as Alice to ``"B"``,
+and successfully did so in between ``Alice``'s two requests.  Then ``Alice``'s
+request will implicitly overwrite ``Bob``s ``secret`` and it will be lost forever.
+
+Even thought the example is fairly trivial example it illustrates pretty clear
+that idempotency does not ensure safe updates in a concurrent environment.
+Also, consider how such API semantics in a large distributed environment can
+cause a a lot of data loss and race conditions if not properly implemented.
+
+### Avoiding race conditions
+One approach to avoid such race conditions is to implement the API with
+[Optimistic Locking](http://stackoverflow.com/questions/129329/optimistic-vs-pessimistic-locking)
+semantics. We can do that by introducing a version number, or other methods such
+as an hash or timestamps, to the data model. I prefer using version numbers because
+they are easy to understand and update. I believe timestamps to be a bad choice
+ because [clocks are hard to keep synchronized](https://www.google.no/url?sa=t&rct=j&q=&esrc=s&source=web&cd=2&cad=rja&uact=8&ved=0ahUKEwjog-O0q5PLAhUMIpoKHc9IC6YQFggnMAE&url=http%3A%2F%2Famturing.acm.org%2Fp558-lamport.pdf&usg=AFQjCNGDBbwvlZwKdG6k44j2O0gIOoEi9w).
```

Comment:
I guess timestamp can just be dropped, as it can lead to byzantine errors. Just don't mention it in the solution space.